exaudfclient/docs/script_options_design.md

```diff
@@ -0,0 +1,457 @@
+# Design Document
+
+This document details the design aspects of the new Exasol UDF Client Script Options parser, based on the high-level requirements outlined in the System Requirement Specification.
+
+## Acknowledgments
+
+This document's section structure is derived from the "[arc42](https://arc42.org/)" architectural template by Dr. Gernot Starke, Dr. Peter Hruschka.
+
+## Constraints
+
+- The parser implementation must be in C++.
+- The chosen parser implementation is [ctpg](https://github.com/peter-winter/ctpg), which supports the definition of Lexer and Parser Rules in C++ code.
+- The selected parser should allow easy encapsulation in a custom C++ namespace (UDF client linker namespace constraint)
+- The selected parser should not depend on additional runtime dependencies
+- The selected parser should have minimal compile time dependencies, i.e. no additional shared libraries or tools to generate C++ code
+
+### Requirement Overview
+
+Please refer to the [System Requirement Specification](script_options_requirments.md) for user-level requirements.
+
+## Building Blocks
+
+### Overall Architecture
+
+#### Component Overview
+
+![Components](diagrams/OveralScriptOptionalsBuildingBlocks.png)
+
+At the very high level there can be distinguished between the generic "Script Options Parser" module which parses a UDF script code and returns the found script options, and the "Script Options Parser Handler" which converts the Java UDF specific script options. In both modules there are specific implementation for the legacy parser and the new CTPG based parser.
+
+### Script Options Parser
+
+The parser component can be used to parse any script code (Java, Python, R) for any script options. It provides simplistic interfaces, which are different between the two versions, which accept the script code as input and return the found script option(s). 
+
+#### Legacy Parser
+
+The legacy parser (V1) parser searches for one specific script option. The parser starts from beginning of the script code. If found, the parser immediately removes the script option from the script code and returns the option value. It validates the 
```

Comment:
In a narrow interpretation of arc42, the removal would be part of the runtime behavior and not the building blocks.